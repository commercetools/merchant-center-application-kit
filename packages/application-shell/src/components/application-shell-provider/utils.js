import { getSupportedLocale } from '@commercetools-frontend/l10n';

// eslint-disable-next-line import/prefer-default-export
export const getBrowserLocale = window => {
  const browserLocale = window && window.navigator && window.navigator.language;
  return getSupportedLocale(browserLocale);
};
