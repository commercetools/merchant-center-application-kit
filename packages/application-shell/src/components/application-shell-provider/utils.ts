import { getSupportedLocale } from '@commercetools-frontend/l10n';

// eslint-disable-next-line import/prefer-default-export
export const getBrowserLocale = (win: Window) => {
  const browserLocale = win && win.navigator && win.navigator.language;
  return getSupportedLocale(browserLocale);
};
