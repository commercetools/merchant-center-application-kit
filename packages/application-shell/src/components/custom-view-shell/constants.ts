// TODO: This file is duplicated (also exists in application-components custom views components)
export const CUSTOM_VIEWS_EVENTS_NAMES = {
  CUSTOM_VIEW_BOOTSTRAP: 'custom-view-bootstrap',
  CUSTOM_VIEW_INITIALIZATION: 'custom-view-initialization',
};

export const CUSTOM_VIEWS_EVENTS_META = {
  SOURCE: 'mc-host-application',
  DESTINATION_PREFIX: 'custom-view-',
};
