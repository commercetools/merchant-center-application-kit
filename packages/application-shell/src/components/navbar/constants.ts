export const NAVBAR = {
  expanderHeight: '50px',
  expanderSize: '40px',
  iconSize: '24px',
  iconSizeHover: '28px',
  itemSize: '48px',
  itemHeight: '56px',
  sublistIndentationWhenCollapsed: '72px',
  sublistIndentationWhenExpanded: '248px',
  sublistWidth: '253px',
  leftNavigationTransition: 'all 150ms cubic-bezier(1, 0, 0.58, 1)',
  widthLeftNavigation: '80px',
  widthLeftNavigationWhenExpanded: '256px',
} as const;
