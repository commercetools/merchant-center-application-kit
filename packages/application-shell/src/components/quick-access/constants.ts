// eslint-disable-next-line import/prefer-default-export

export const permissions = {
  ViewOrders: 'ViewOrders',
  ManageOrders: 'ManageOrders',
  ViewProducts: 'ViewProducts',
  ManageProducts: 'ManageProducts',
  ViewCategories: 'ViewCategories',
  ManageCategories: 'ManageCategories',
  ViewCustomers: 'ViewCustomers',
  ManageCustomers: 'ManageCustomers',
  ViewCustomerGroups: 'ViewCustomerGroups',
  ManageCustomerGroups: 'ManageCustomerGroups',
  ViewProductDiscounts: 'ViewProductDiscounts',
  ManageProductDiscounts: 'ManageProductDiscounts',
  ViewDiscountCodes: 'ViewDiscountCodes',
  ManageDiscountCodes: 'ManageDiscountCodes',
  ViewCartDiscounts: 'ViewCartDiscounts',
  ManageCartDiscounts: 'ManageCartDiscounts',
  ViewProjectSettings: 'ViewProjectSettings',
  ManageProjectSettings: 'ManageProjectSettings',
  ViewDeveloperSettings: 'ViewDeveloperSettings',
  ManageDeveloperSettings: 'ManageDeveloperSettings',
  ViewProductTypes: 'ViewProductTypes',
  ManageProductTypes: 'ManageProductTypes',
} as const;
