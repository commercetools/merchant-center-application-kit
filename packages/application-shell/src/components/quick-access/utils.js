import flatMap from 'lodash.flatmap';

// copied from @commercetools-local/utils/query-string
export const sanitize = param => {
  if (typeof param === 'string')
    return (
      param
        // Replace all \ with \\ (to prevent generate escape characters)
        .replace(/\\/g, '\\\\')
        // Replace all " with \"
        .replace(/"/g, '\\"')
    );

  return param;
};

export const flattenResults = results =>
  flatMap(results, result =>
    result.subCommands
      ? [result, ...flattenResults(result.subCommands)]
      : result
  );

// Once ui-kit exposes its fallback mechanism, we can use the same one here
export const translate = (nameAllLocales, projectDataLocale) => {
  const matchedTranslation = nameAllLocales.find(
    translation => translation.locale === projectDataLocale && translation.value
  );
  if (matchedTranslation) return matchedTranslation.value;

  const fallback = nameAllLocales.find(translation => translation.value);
  return fallback.value;
};
