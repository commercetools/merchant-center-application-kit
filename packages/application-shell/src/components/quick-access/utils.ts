import type { TLocalizedString } from '../../types/generated/ctp';
import type { Command, ExecGraphQlQuery } from './types';

export const sanitize = (param: string) =>
  param
    // Replace all \ with \\ (to prevent generate escape characters)
    .replace(/\\/g, '\\\\')
    // Replace all " with \"
    .replace(/"/g, '\\"');

export const flattenCommands = async (
  results: Command[],
  execQuery: ExecGraphQlQuery
) => {
  async function flatten(commands: Command[]): Promise<Command[]> {
    return commands.reduce(async (prevPromise: Promise<Command[]>, command) => {
      const prevResults = await prevPromise;
      if (command.subCommands) {
        if (typeof command.subCommands === 'function') {
          const subCommands = await command.subCommands(execQuery);
          const flattenSubCommands = await flatten(subCommands);
          return [...prevResults, command, ...flattenSubCommands];
        }
        const flattenSubCommands = await flatten(command.subCommands);
        return [...prevResults, command, ...flattenSubCommands];
      }
      return [...prevResults, command];
    }, Promise.resolve([]));
  }

  return await flatten(results);
};

// Once ui-kit exposes its fallback mechanism, we can use the same one here
export const translate = (
  nameAllLocales: TLocalizedString[],
  projectDataLocale: string
) => {
  const matchedTranslation = nameAllLocales.find(
    (translation) =>
      translation.locale === projectDataLocale && translation.value
  );
  if (matchedTranslation) return matchedTranslation.value;

  // Fall back to the first available locale
  if (nameAllLocales.length > 0) return nameAllLocales[0].value;
  return '';
};
