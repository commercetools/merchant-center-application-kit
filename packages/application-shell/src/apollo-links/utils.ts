import type { GraphQLError } from 'graphql';
import type { ErrorResponse } from 'apollo-link-error';
import type { ServerError, ServerParseError } from 'apollo-link-http-common';
import type { TApolloContext } from '../utils/apollo-context';

import { GRAPHQL_TARGETS } from '@commercetools-frontend/constants';
import { SUPPORTED_HEADERS } from '../constants';

type TTokenRetryGraphQlTarget =
  | typeof GRAPHQL_TARGETS.COMMERCETOOLS_PLATFORM
  | typeof GRAPHQL_TARGETS.SETTINGS_SERVICE
  | typeof GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND
  | typeof GRAPHQL_TARGETS.ADMINISTRATION_SERVICE;

const getSkipTokenRetry = (context: TApolloContext): boolean => {
  const skipTokenRetry = Boolean(context.skipTokenRetry);

  return skipTokenRetry;
};

const forwardTokenRetryHeader = (
  headers: TApolloContext['headers']
): TApolloContext['headers'] => ({
  ...headers,
  [SUPPORTED_HEADERS.X_TOKEN_RETRY]: 'true',
});

// This link retries requests to the CTP API that have been rejected
// because of an invalid/expired oauth token.
// To do so, we resend the request with the header "X-Force-Token: true"
// so that the MC BE can issue a new token.
// NOTE: the retry is not meant to work for the MC access token.
const getDoesGraphQLTargetSupportTokenRetry = (
  context: TApolloContext
): boolean => {
  const graphQLTarget = (context.headers?.[
    SUPPORTED_HEADERS.X_GRAPHQL_TARGET
  ] || context.headers?.[SUPPORTED_HEADERS.X_GRAPHQL_TARGET.toLowerCase()]) as
    | TTokenRetryGraphQlTarget
    | undefined;

  return Boolean(
    graphQLTarget &&
      [
        GRAPHQL_TARGETS.COMMERCETOOLS_PLATFORM,
        GRAPHQL_TARGETS.ADMINISTRATION_SERVICE,
        GRAPHQL_TARGETS.SETTINGS_SERVICE,
        GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND,
      ].includes(graphQLTarget)
  );
};

const isHttpError = (
  error: ErrorResponse['networkError']
): error is ServerError | ServerParseError =>
  (error as ServerError).statusCode !== undefined ||
  (error as ServerParseError).statusCode !== undefined;

const isGraphQLError = (
  error: ErrorResponse['graphQLErrors']
): error is GraphQLError[] =>
  Array.isArray(error) &&
  error.some(
    (err) =>
      (err as GraphQLError).message !== undefined ||
      (err as GraphQLError).extensions !== undefined
  );

export {
  getSkipTokenRetry,
  forwardTokenRetryHeader,
  getDoesGraphQLTargetSupportTokenRetry,
  isHttpError,
  isGraphQLError,
};
