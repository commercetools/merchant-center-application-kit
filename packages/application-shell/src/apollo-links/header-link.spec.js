import { ApolloLink, execute, Observable } from 'apollo-link';
import gql from 'graphql-tag';
import waitFor from 'wait-for-observables';
import { GRAPHQL_TARGETS } from '@commercetools-frontend/constants';
import { getCorrelationId } from '../utils';
import headerLink from './header-link';

jest.mock('../utils/', () => ({
  getCorrelationId: jest.fn(() => 'test-correlation-id'),
  selectProjectKeyFromUrl: jest.fn(() => 'project-1'),
  selectUserId: jest.fn(() => 'user-1'),
}));

describe('headerLink', () => {
  it('should be an instance of ApolloLink', () => {
    expect(headerLink).toBeInstanceOf(ApolloLink);
  });
});

const query = gql`
  {
    sample {
      id
    }
  }
`;

describe('with valid target', () => {
  let context;
  let debugLink;
  let link;
  let terminatingLinkStub;

  beforeEach(async () => {
    debugLink = new ApolloLink((operation, forward) => {
      context = operation.getContext();

      return forward(operation);
    });

    terminatingLinkStub = jest.fn(() => Observable.of({}));

    link = ApolloLink.from([headerLink, debugLink, terminatingLinkStub]);

    await waitFor(
      execute(link, {
        query,
        variables: { target: GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND },
      })
    );
  });

  it('should set headers matching snapshot', () => {
    expect(context).toMatchSnapshot();
  });

  it('should set `X-Project-Key`-Header', () => {
    expect(context.headers).toEqual(
      expect.objectContaining({
        'X-Project-Key': 'project-1',
      })
    );
  });

  it('should set `X-Graphql-Target`-Header', () => {
    expect(context.headers).toEqual(
      expect.objectContaining({
        'X-Graphql-Target': GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND,
      })
    );
  });

  it('should include `mc` in the `X-Correlation-Id`-Header', () => {
    expect(context.headers).toEqual(
      expect.objectContaining({
        'X-Correlation-Id': 'test-correlation-id',
      })
    );
  });

  it('should pass "userId" param to "getCorrelationId"', () => {
    expect(getCorrelationId).toHaveBeenCalledWith({ userId: 'user-1' });
  });

  describe('with project key in variables', () => {
    const projectKey = 'test-project-key';

    beforeEach(async () => {
      await waitFor(
        execute(link, {
          query,
          variables: {
            target: GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND,
            projectKey,
          },
        })
      );
    });

    it('should set headers matching snapshot', () => {
      expect(context).toMatchSnapshot();
    });

    it('should set `X-Project-Key`-Header', () => {
      expect(context.headers).toEqual(
        expect.objectContaining({
          'X-Project-Key': projectKey,
        })
      );
    });
  });
});
