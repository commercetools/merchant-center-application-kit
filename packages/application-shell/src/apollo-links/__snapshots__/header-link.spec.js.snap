// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with valid target should set headers matching snapshot 1`] = `
Object {
  "headers": Object {
    "X-Correlation-Id": "test-correlation-id",
    "X-Graphql-Target": "mc",
    "X-Project-Key": "project-1",
  },
}
`;

exports[`with valid target with project key in variables should set headers matching snapshot 1`] = `
Object {
  "headers": Object {
    "X-Correlation-Id": "test-correlation-id",
    "X-Graphql-Target": "mc",
    "X-Project-Key": "test-project-key",
  },
}
`;
